### ISSUE TITLE
Connection refused

### ISSUE THREAD

**Issue by MatteoGioioso** [wmnnd/nginx-certbot#11]
I know it is a duplicate of #10 .
But what "described" in that issue looks like not a valid solution.
Anyway this is the error

```
The following errors were reported by the server:

   Domain: mydomain.org
   Type:   connection
   Detail: Fetching
   http://mydomain.org/.well-known/acme-challenge/y1TyeXK2Il0hfnqROt-3Arn_6Oq28ffL4kshJZ50pic:
   Connection refused
```

this is all the log
```
matteo@domain:~/domain$ ./init-letsencrypth.sh
Existing data found for domain.com. Continue and replace existing certificate? (y/N) n
matteo@domain:~/domain$ ./init-letsencrypth.sh
Existing data found for domain.com. Continue and replace existing certificate? (y/N) y
### Creating dummy certificate for domain.com ...
Creating network "domain_default" with the default driver
Generating a RSA private key
..+++++
.............+++++
writing new private key to '/etc/letsencrypt/live/domain.com/privkey.pem'
req: Can't open "/etc/letsencrypt/live/domain.com/privkey.pem" for writing, No such file or directory

### Starting nginx ...
Creating domain ... done

### Deleting dummy certificate for domain.com ...

### Requesting Let's Encrypt certificate for domain.com ...
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator webroot, Installer None
Obtaining a new certificate
Performing the following challenges:
http-01 challenge for admin.domain.com
http-01 challenge for domain.com
http-01 challenge for www.admin.domain.com
http-01 challenge for www.domain.com
Using the webroot path /var/www/certbot for all unmatched domains.
Waiting for verification...
Cleaning up challenges
Failed authorization procedure. www.domain.com (http-01): urn:ietf:params:acme:error:connection :: The server could not connect to the client to verify the domain ...

IMPORTANT NOTES:
 - The following errors were reported by the server:

   Domain: www.domain.com
   Type:   connection
   Detail: Fetching
   http://www.domain.com/.well-known/acme-challenge/CksLQlZjt7xYEf683LHRV87PRAnF9KglD3Yhpw9DESk:
   Connection refused

   Domain: admin.domain.com
   Type:   connection
   Detail: Fetching
   http://admin.domain.com/.well-known/acme-challenge/o6XfSqjxuzupQiiQ_DJ-mUUWmMTgZvyLZKFrZ1RAd7s:
   Connection refused

   Domain: www.admin.domain.com
   Type:   connection
   Detail: Fetching
   http://www.admin.domain.com/.well-known/acme-challenge/QmlFWKvP8L_OE3R5tI0BPG3fjcOwodFpNOT0beNJitE:
   Connection refused

   Domain: domain.com
   Type:   connection
   Detail: Fetching
   http://domain.com/.well-known/acme-challenge/R3RlQ30GFfQQ4yck1l3sA252w-mLy2tNo0DJa_s_wvc:
   Connection refused

   To fix these errors, please make sure that your domain name was
   entered correctly and the DNS A/AAAA record(s) for that domain
   contain(s) the right IP address. Additionally, please check that
   your computer has a publicly routable IP address and that no
   firewalls are preventing the server from communicating with the
   client. If you're using the webroot plugin, you should also verify
   that you are serving files from the webroot path you provided.

### Reloading nginx ...
2019/03/27 02:55:14 [notice] 7#7: signal process started
```

I have notice this, maybe it is the source of the error
```
req: Can't open "/etc/letsencrypt/live/domain.com/privkey.pem" for writing, No such file or directory
```

If you need more info I could give it to you.
Thanks

**Comment by MatteoGioioso** [comment]
I have investigated a bit and I found what could be a potential bottleneck for your script and correct me if I am wrong:
when you execute this command
`docker-compose up --force-recreate -d nginx`
you are supposing  that nginx does not have any dependency service.
In my case nginx was **not starting** successfully because my `upstream` was connected to another container; to solve this I had to modify the command to start every services ```docker-compose up --force-recreate -d```

**Comment by MatteoGioioso** [comment]
ok, finally I made it work by modifying the script
so 2 points:
- the script does not take into account for additional services
- the script does not work for multiple domains

```
#!/bin/bash

domains=(domain.com www.domain.com admin.domain.com www.admin.domain.com)
rsa_key_size=4096
data_path="./certbot"
email="someemail@gmail.com" # Adding a valid address is strongly recommended
staging=0 # Set to 1 if you're testing your setup to avoid hitting request limits

if [ -d "$data_path" ]; then
  read -p "Existing data found for $domains. Continue and replace existing certificate? (y/N) " decision
  if [ "$decision" != "Y" ] && [ "$decision" != "y" ]; then
    exit
  fi
fi


if [ ! -e "$data_path/conf/options-ssl-nginx.conf" ] || [ ! -e "$data_path/conf/ssl-dhparams.pem" ]; then
  echo "### Downloading recommended TLS parameters ..."
  mkdir -p "$data_path/conf"
  curl -s https://raw.githubusercontent.com/certbot/certbot/master/certbot-nginx/certbot_nginx/options-ssl-nginx.conf > "$data_path/conf/options-ssl-nginx.conf"
  curl -s https://raw.githubusercontent.com/certbot/certbot/master/certbot/ssl-dhparams.pem > "$data_path/conf/ssl-dhparams.pem"
  echo
fi

for domain in "${domains[@]}"; do
  echo "### Removing old certificate for $domain ..."
  docker-compose run --rm --entrypoint "\
    rm -Rf /etc/letsencrypt/live/$domain && \
    rm -Rf /etc/letsencrypt/archive/$domain && \
    rm -Rf /etc/letsencrypt/renewal/$domain.conf" certbot
  echo
done

for domain in "${domains[@]}"; do
  echo "### Creating dummy certificate for $domain ..."
  path="/etc/letsencrypt/live/$domain"
  mkdir -p "$data_path/conf/live/$domain"
  docker-compose run --rm --entrypoint "\
    openssl req -x509 -nodes -newkey rsa:1024 -days 1\
      -keyout "$path/privkey.pem" \
      -out "$path/fullchain.pem" \
      -subj '/CN=localhost'" certbot
  echo
done

echo "### Starting nginx ..."
docker-compose up --force-recreate -d
echo

for domain in "${domains[@]}"; do
  echo "### Removing dummy certificate for $domain ..."
  docker-compose run --rm --entrypoint "\
    rm -Rf /etc/letsencrypt/live/$domain" certbot
  echo
done

echo "### Requesting Let's Encrypt certificates ..."

# Select appropriate email arg
case "$email" in
  "") email_arg="--register-unsafely-without-email" ;;
  *) email_arg="--email $email" ;;
esac

# Enable staging mode if needed
if [ $staging != "0" ]; then staging_arg="--staging"; fi

for domain in "${domains[@]}"; do
  docker-compose run --rm --entrypoint "\
    certbot certonly --webroot -w /var/www/certbot \
      $staging_arg \
      $email_arg \
      -d $domain \
      --rsa-key-size $rsa_key_size \
      --agree-tos \
      --force-renewal" certbot
  echo
done

echo "### Reloading nginx ..."
docker-compose exec nginx nginx -s reload
```

**Comment by tmadison-gpsw** [comment]
I am seeing the same issue on a previously working setup. I started from scratch and I am still getting ```
Domain: www.domain.com
Type:   connection
Detail: Fetching
http://www.domain.com/.well-known/acme- challenge/CksLQlZjt7xYEf683LHRV87PRAnF9KglD3Yhpw9DESk:
Connection refused
```

**Comment by MatteoGioioso** [comment]
@tmadison-gpsw did you check if your nginx server started correctly?

**Comment by tmadison-gpsw** [comment]
@MatteoGioioso My bad I had added additional redirects to the original `app.conf`, I got it working now, thanks!

**Comment by wmnnd** [comment]
Glad to hear it’ſ working now!

**Comment by YoniSchirris** [comment]
I'm still getting similar errors. 
- I have no more "example.org" in any of the files
- I'm only testing for a single domain pointing to a static IP on a linux EC2 server where I run docker-compose
- I've tried the exact steps in the medium tutorial, and I've taken the latest code from the github repo.
- I've tried it with the proposed nginx server, and with my own set-up of my server which, if run by itself, works (but only over HTTP)

I have the feeling the server with docker-compose doesn't start correctly. But I have no idea how to check this or how to fix it
